src/gtk/tests/widgets/test_table.py

```diff
@@ -0,0 +1,237 @@
+
+        # Make sure the row got stored correctly
+        result_row = self.gtk_table.store[path]
+        self.assertEqual(tuple(result_row), row_data)
+
+    def test_remove(self):
+        # Insert a row
+        row = self.table.data.insert(0, "1", "2")
+
+        # Make sure it's in there
+        self.assertTrue(row in self.gtk_table.rows)
+
+        # Then remove it
+        self.gtk_table.remove(row)
+
+        # Make sure its gone
+        self.assertFalse(row in self.gtk_table.rows)
+
+    def test_change(self):
+        # Insert a row
+        row = self.table.data.insert(0, "1", "2")
+
+        # Make sure it's in there
+        self.assertTrue(row in self.gtk_table.rows)
+
+        # Change a column
+        row.one = "something_changed"
+        # (not testing that self.gtk_table.change is called. The Core API
+        # unit tests should ensure this already.)
+
+        # Get the Gtk.TreeIter
+        tree_iter = self.gtk_table.rows[row]
+
+        # Make sure it's a Gtk.TreeIter
+        self.assertTrue(isinstance(tree_iter, Gtk.TreeIter))
+
+        # Make sure it's the correct Gtk.TreeIter
+        self.assertEqual(row, self.gtk_table.get_row(tree_iter))
+
+        # Make sure the value changed
+        path = self.gtk_table.store.get_path(tree_iter)
+        result_row = self.gtk_table.store[path]
+        self.assertEqual(tuple(result_row), (row.one, row.two))
+
+    def test_method_compare_tree_iters(self):
+        store = self.gtk_table.store
+        store.clear()
+        store.append(("1", "2"))
+
+        # Gtk.TreeIters can't be compared directly
+        self.assertNotEqual(store[0].iter, store[0].iter)
+
+        self.assertTrue(self.gtk_table.compare_tree_iters(
+            store[0].iter,
+            store[0].iter
+        ))
+
+    def test_method_get_row(self):
+        # Put data in the Table
+        self.table.table = []
+        A = self.table.data.append(None, one="A1", two="A2")
+        tree_iter = self.gtk_table.rows[A]
+        self.assertEqual(A, self.gtk_table.get_row(tree_iter))
+
+        with self.assertRaises(TypeError):
+            self.gtk_table.get_row(None)
+
+    def test_row_persisence(self):
+        A = self.table.data.insert(0, one="A1", two="A2")
+        B = self.table.data.insert(0, one="B1", two="B2")
+
+        # B should now precede A
+        # tests passes if A "knows" it has moved to index 1
+
+        self.assertEqual(tuple(self.gtk_table.store[0]), ("B1", "B2"))
+        self.assertEqual(tuple(self.gtk_table.store[1]), ("A1", "A2"))
+
+    def test_on_select_root_row(self):
+        # Insert two dummy rows
+        self.table.data = []
+        A = self.table.data.append(None, one="A1", two="A2")
+        B = self.table.data.append(None, one="B1", two="B2")
+
+        # Create a flag
+        succeed = False
+
+        def on_select(table, row, *kw):
+            # Make sure the right row was selected
+            self.assertEqual(row, B)
+
+            nonlocal succeed
```

Comment:
TIL about nonlocal. It's a bit hard to follow exactly _which_ `succeed` is being used in the assert.

Reply:
Agreed; under normal circumstances (i.e. not using GTK event loop), I would just have the handler raise an `Succeed` Exception and then `self.assertRaises(Succeed)`, but exceptions get raised on the event loop and seem to pass unnoticed by the test runner. I am certainly open to suggestions for a more conventional or readable method.